src/quic/preferredaddress.cc

```diff
@@ -0,0 +1,147 @@
+#include "preferredaddress.h"
+#include <env-inl.h>
+#include <node_errors.h>
+#include <node_sockaddr-inl.h>
+#include <util-inl.h>
+#include <ngtcp2/ngtcp2.h>
+#include <uv.h>
+#include <v8.h>
+
+namespace node {
+
+using v8::Just;
+using v8::Local;
+using v8::Maybe;
+using v8::Nothing;
+using v8::Value;
+
+namespace quic {
+
+namespace {
+template <int FAMILY>
+std::optional<const PreferredAddress::AddressInfo> get_address_info(
+    const ngtcp2_preferred_addr& paddr) {
+  if constexpr (FAMILY == AF_INET) {
+    if (!paddr.ipv4_present) return std::nullopt;
+    PreferredAddress::AddressInfo address;
+    address.family = FAMILY;
+    address.port = paddr.ipv4_port;
+    if (uv_inet_ntop(FAMILY, paddr.ipv4_addr,
+                     address.host, sizeof(address.host)) == 0) {
+      address.address = address.host;
+    }
+    return address;
+  } else {
+    if (!paddr.ipv6_present) return std::nullopt;
+    PreferredAddress::AddressInfo address;
+    address.family = FAMILY;
+    address.port = paddr.ipv6_port;
+    if (uv_inet_ntop(FAMILY, paddr.ipv6_addr,
+                     address.host, sizeof(address.host)) == 0) {
+      address.address = address.host;
+    }
+    return address;
+  }
+}
+
+template <int FAMILY>
+void copy_to_transport_params(
+    ngtcp2_transport_params* params,
+    const sockaddr* addr) {
+  params->preferred_address_present = true;
+  if constexpr (FAMILY == AF_INET) {
+    const sockaddr_in* src = reinterpret_cast<const sockaddr_in*>(addr);
+    params->preferred_address.ipv4_port = SocketAddress::GetPort(addr);
+    memcpy(params->preferred_address.ipv4_addr,
+           &src->sin_addr,
+           sizeof(params->preferred_address.ipv4_addr));
+  } else {
+    DCHECK_EQ(FAMILY, AF_INET6);
+    const sockaddr_in6* src = reinterpret_cast<const sockaddr_in6*>(addr);
+    params->preferred_address.ipv6_port = SocketAddress::GetPort(addr);
+    memcpy(params->preferred_address.ipv6_addr,
+           &src->sin6_addr,
+           sizeof(params->preferred_address.ipv4_addr));
+  }
+  UNREACHABLE();
+}
+
+bool resolve(const PreferredAddress::AddressInfo& address,
+             uv_getaddrinfo_t* req) {
+  addrinfo hints{};
+  hints.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+  hints.ai_family = address.family;
+  hints.ai_socktype = SOCK_DGRAM;
+
+  // ngtcp2 requires the selection of the preferred address
+  // to be synchronous, which means we have to do a sync resolve
+  // using uv_getaddrinfo here.
+  return uv_getaddrinfo(nullptr,
+                        req,
+                        nullptr,
+                        address.host,
+                        std::to_string(address.port).c_str(),
```

Comment:
Nit: Depending on the usage & execution count, this line might be super-costly.

Reply:
I'll add a comment here marking it as a future perf optimization